### PR TITLE
Explore: Adds ability to go back to dashboard, optionally with query changes

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Allows a user to save the query associated with a dashboard panel from within Explore if the user came to Explore through the panel.
This has been implemented according to the MVP specs detailed in the design document.

### CODE REVIEW

File: public/app/core/utils/explore.ts [hunk 1]
```diff
@@ -106,6 +106,7 @@ export async function getExploreUrl(
         ...state,
         datasource: panelDatasource.name,
         queries: exploreTargets.map(t => ({ ...t, datasource: panelDatasource.name })),
+        originPanel: panel.id,
```
Comment (davkal) [inline public/app/core/utils/explore.ts]: I never reach this code when using this with a Prometheus datasource. It's using the `if` branch above.

**Comment by torkelo** [comment]
Sorry, I feel that the scenario of saving the origin dashboard from Explore is wrong. Sorry should have voiced this more clearly in the design doc, did not know this had moved passed design doc phase, not seen of much demand for this. And maybe missed this solution proposal (to save directly from Explore). 

This feels very wrong to me, maybe if explore was just a different UI around the same visualization & settings. But right now it is very different and there is no guarantee that what you see in explore is the same you will see in the Dashboard. Saving directly from explore feels even worse when you do not even have the context of the panel. 

Maybe copy the query, or go back to the panel you came from (but in edit mode) to see the changed query in that context.

**Comment by kaydelaney** [comment]
> A safer option would be to go back to the panel you came from (but in edit mode) to see the changed query in that correct context and what effect it has on the panel you came from. The you can save the dashboard using the existing save workflows.

To this effect, should we add URL parameters representing the state of the panel's query? Not sure if there would be another way to change the panel editor's state.

**Comment by jessover9000** [comment]
Reviewing the editing/saving mechanism, I felt like all in all this adjustment works fine. Good work, Kay and everyone else! There is one detail which I think could be improved: I found that the system could give more visual feedback on which queries users change. I did understand that after changing something in explore and going back to the panel, the change showed up in my queries. but I have only had examples with very few queries, therefore I'm wondering if it might hard to see that you have unsaved changed queries in your list depending on the usecase. I propose a "- modified" flag for the query list item that was edited in explore. I have made a quick sketch in Figma: https://www.figma.com/file/QJtNKheEy5udxr7CG03IiN4G/Explore-Export-query-to-dashboard-flow-chart?node-id=0%3A1

File: public/app/features/dashboard/services/DashboardSrv.ts [hunk 2]
```diff
@@ -37,54 +44,51 @@ export class DashboardSrv {
...
-    if (options.edit) {
-      urlParams.edit = true;
-    } else {
-      delete urlParams.edit;
-      delete urlParams.tab;
+    // I hate using these truthy converters (!!) but in this case
+    // I think it's appropriate. edit can be null/false/undefined and
+    // here i want all of those to compare the same
+    if (fullscreen === urlParams.fullscreen && edit === !!urlParams.edit) {
+      ['fullscreen', 'edit', 'panelId', 'tab'].forEach(key => {
```
Comment (torkelo) [inline public/app/features/dashboard/services/DashboardSrv.ts]: nice refactoring of this messy function! :+1: 

Instead of the forEach use regular for loop with for of

**Review by torkelo — COMMENTED** [review]
Great work, looks promising! 

Some code remains from the previous solution (changes to DashboardSrv & SaveDashboardModalCtrl), looks like those changes are no longer used? 

Added some questions & code comments.

File: public/app/core/services/backend_srv.ts [hunk 3]
```diff
@@ -263,14 +263,15 @@ export class BackendSrv implements BackendService {
...
+    dash: DashboardModel,
+    { message = '', folderId, overwrite }: { message?: string; folderId?: number; overwrite?: boolean } = {}
+  ) {
     return this.post('/api/dashboards/db/', {
       dashboard: dash,
-      folderId: options.folderId,
-      overwrite: options.overwrite === true,
-      message: options.message || '',
+      folderId,
+      overwrite,
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/core/services/backend_srv.ts]: ah, you're right. should be fixed if i set `overwrite = false` as a default parameter.

File: public/app/features/dashboard/components/SaveModals/SaveDashboardModalCtrl.ts [hunk 4]
```diff
@@ -105,11 +107,12 @@ export class SaveDashboardModalCtrl {
     const saveModel = dashboard.getSaveModelClone(options);
 
     this.isSaving = true;
-
-    return this.dashboardSrv.save(saveModel, options).then(this.postSave.bind(this, options));
+    return this.dashboardSrv
+      .save(saveModel, { ...options, redirect: this.redirect })
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/dashboard/components/SaveModals/SaveDashboardModalCtrl.ts]: Sure, I can remove them.

File: public/app/features/explore/ExploreToolbar.tsx [hunk 5]
```diff
@@ -112,6 +118,25 @@ export class UnConnectedExploreToolbar extends PureComponent<Props, {}> {
...
+
+    const titleSlug = kbn.slugifyForUrl(dash.title);
+    getLocationSrv().update({
+      path: `/d/${dash.uid}/:${titleSlug}`,
+      query: {
+        fullscreen: true,
+        edit: true,
+        panelId: originPanel,
+        fromExplore: true,
+        queryRows: queries.length,
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/explore/ExploreToolbar.tsx]: it's the number of queries to take from explore history in local storage. Needed since otherwise there would be no way to know if the first, say, 3 entries in the `grafana.explore.history[dataSource]` array refer to 3 query rows at one point in history, or 1 query row at 3 points in history.
I could change the name to `queryRowCount` if that makes it more clear?

**Comment by davkal** [comment]
This is great to play around with, after having used it I have some more UX feedback:

- I find the return-to-origin-dashboard useful, I'd like to make this a separate (primary) action
- return-to-origin-dashboard should always be possible after a jump (e.g, even after datasource change, but may be "forgotten" once Explore has been split)
- return-to-origin-dashboard-with-changes IMO is a secondary action to the former, so could either piggy-backs (via showing a modal that confirms/rejects "The queries have been modified. Would you like to update the dashboard panel?") or becomes its own secondary option (e.g., via dropdown item "Return to and update dashboard panel" of the primary action of returning).
- The "modified" signifiers that @jessover9000 suggested would add more modeless feedback about which parts actually changed
- The "modified" signifiers could even carry additional options to "reset" to the original in case some row imports were accidental. 

This is shaping up nicely, and I'm quite happy about how this improves the flow back to panels.

File: public/app/features/dashboard/panel_editor/QueriesTab.tsx [hunk 6]
```diff
@@ -58,10 +63,37 @@ export class QueriesTab extends PureComponent<Props, State> {
...
 
   componentDidMount() {
-    const { panel } = this.props;
+    const { panel, fromExplore, queryRows } = this.props;
     const queryRunner = panel.getQueryRunner();
 
     this.querySubscription = queryRunner.subscribe(this.panelDataObserver, PanelQueryRunnerFormat.both);
+
+    if (fromExplore) {
+      const datasourceType = this.datasources.filter(datasource => datasource.value === panel.datasource)[0].meta.id;
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/dashboard/panel_editor/QueriesTab.tsx]: would something like this work?
```typescript
  private get datasourceType(): string | undefined {
    try {
      return this.datasources.filter(datasource => datasource.value === this.props.panel.datasource)[0].meta.id;
    } catch {
      return undefined;
    }
  }
```

**Comment by jessover9000** [comment]
Research suggests that "modified" is not clear UX writing-wise, so I suggest we change to "edited". A newly created and unsaved query could also be labeled "edited". @kaydelaney

File: public/app/features/dashboard/panel_editor/QueriesTab.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard/panel_editor/QueriesTab.tsx]: I was referring this this whole if case, and process of taking the query from explore and modifying the current dashboard model. That could also be done in the dashboard init phase. Or in seperate function as it's not really the responsibility of the QueriesTab ui component I feel.

File: public/app/features/dashboard/components/SaveModals/SaveDashboardModalCtrl.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard/components/SaveModals/SaveDashboardModalCtrl.ts]: great

File: public/app/features/explore/ExploreToolbar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/explore/ExploreToolbar.tsx]: yes that would help, as the queryRows name to me makes this variable/property represent an array of rows so the Count suffix helps make it clear it's just a number. 

So what was the reasoning behind using localstorage instead of just storing the modified query in the explore redux state? this state should not need to survive full page reloads. But maybe it aids developing and testing this feature?

File: public/app/features/dashboard/panel_editor/QueriesTab.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/dashboard/panel_editor/QueriesTab.tsx]: makes sense, thanks

**Comment by jessover9000** [comment]
@kaydelaney added another view in Figma for a deleted query. The way I imagine it is the accordeon collapses and the background color of the query accordeon fades out, but it is still there with the same measurements/whitespace until you save the change. On undo, the collapse/fade out animation could reverse :)

File: public/app/features/explore/ExploreToolbar.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/explore/ExploreToolbar.tsx]: No reasoning beyond that the query state is already stored there, so I didn't think it'd be necessary to duplicate that state in redux. But if you think it's clearer just to have the modified query in redux I can change it?

[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/explore/ExploreToolbar.tsx]: It would not be duplicate. The dashboard model is stored in redux , and the process of loading and initiation is In redux. So reading this query param and modifying the query during dashboard init is an option to keep this logic out of UI component. Worth investigating

[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/explore/ExploreToolbar.tsx]: I guess by "duplicate" I mean having the modified query both in localstorage (via Explore history), and also having it in redux. I know that while using Explore the queries are also available from redux, but that gets reset when the Explore wrapper gets unmounted (via `resetExploreAction`), so it's no longer accessible during dashboard init. Maybe I'm misunderstanding something though?

[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/explore/ExploreToolbar.tsx]: Ah, I missed the fact that you were using the existing local storage used by the history feature.

> I know that while using Explore the queries are also available from redux, but that gets reset when the Explore wrapper gets unmounted

I was more thinking that when the action "go back to dashboard with updated queries" action is triggered something is stored somewhere in the redux store (like the queries & fromExplore) and then the location change.  

But you could still access the query params and local storage in the dashboard init phase. That process has unit tests as well and would be easier to test I think. 
 
https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/initDashboard.ts

File: public/app/types/dashboard.ts [hunk 7]
```diff
@@ -38,6 +38,7 @@ export interface DashboardMeta {
 }
 
 export interface DashboardDataDTO {
+  [x: string]: any;
```
Comment (davkal) [inline public/app/types/dashboard.ts]: This looks a bit broad. Why is this needed?

File: public/app/features/dashboard/state/initDashboard.ts [hunk 8]
```diff
@@ -198,6 +199,15 @@ export function initDashboard(args: InitDashboardArgs): ThunkResult<void> {
       console.log(err);
     }
 
+    const left = storeState.explore && storeState.explore.left;
+    if (left && left.originPanelId) {
+      const panelArrId = dashboard.panels.findIndex(panel => panel.id === left.originPanelId);
+      if (panelArrId > -1) {
+        dashboard.panels[panelArrId].targets = left.queries.map(query => ({ refId: query.refId, expr: query.expr }));
```
Comment (davkal) [inline public/app/features/dashboard/state/initDashboard.ts]: Not every datasource stores its query in the `expr`. Some even have complex objects. Can we simply copy this new edited object?

File: public/app/features/dashboard/state/initDashboard.ts [hunk 9]
```diff
@@ -198,6 +199,15 @@ export function initDashboard(args: InitDashboardArgs): ThunkResult<void> {
...
+      }
+      dashboard.startRefresh();
+      dispatch(resetExploreAction({ force: true }));
```
Comment (davkal) [inline public/app/features/dashboard/state/initDashboard.ts]: Why do we need to reset Explore here?

File: packages/grafana-ui/src/types/datasource.ts [hunk 10]
```diff
@@ -414,6 +414,8 @@ export interface DataQuery {
   datasource?: string | null;
 
   metric?: any;
+
+  expr?: string;
```
Comment (davkal) [inline packages/grafana-ui/src/types/datasource.ts]: This is a datasource-specific attribute. Can we do without? Also see other comment when copying `refId` and `expr` only.

**Review by davkal — COMMENTED** [review]
Working really well with Prometheus. Now we should make this work for other datasources (use of datasource-specific `expr`), e.g., returning with changes to Graphite loses all changes. 

Another minor thing is that the back-action always leads to the edit mode, but I can also jump to Explore from the dashboard view and the fullscreen view mode. Would be great to have the back-action go back to them. The back-with-changes-action should always go back to the edit view.

File: packages/grafana-ui/src/types/datasource.ts (hunk 10 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/types/datasource.ts]: Ah, I didn't know that. Sure, I'll change this

File: public/app/features/dashboard/state/initDashboard.ts (hunk 9 above)
Comment (kaydelaney) [reply public/app/features/dashboard/state/initDashboard.ts]: So, for now the way I've implemented this feature is that when returning to the origin panel from explore, if we're doing so with changes all the explore state is reset (as is the behavior in master currently) _except_ the queries and the origin panel ID. Then after the dashboard has been initialized with the queries from explore, we completely reset the explore state (with `{ force: true }`), so that on subsequent dashboard loads we aren't taking the modified queries from explore again.

File: public/app/types/dashboard.ts (hunk 7 above)
Comment (kaydelaney) [reply public/app/types/dashboard.ts]: It's definitely broader than it needs to be, but the original type definition doesn't really reflect the structure it's supposed to either. I guess something more accurate would be like:
```ts
{
  annotations: { list: any[] };
  editable: any;
  gnetId: any;
  graphTooltip: number;
  id: number;
  links: any;
  panels: PanelModel[];
  schemaVersion: number;
  style: 'light' | 'dark';
  tags: any;
  templating: { list: any[] };
  time: any;
  timepicker: {
    collapse?: boolean,
    enable?: boolean,
    hidden?: boolean,
    notice?: boolean,
    now?: boolean,
    refresh_intervals?: string[],
    status?: string,
    time_options?: string[],
    type?: string,
  };
  timezone: any;
  title: string;
  uid: string;
  version: number;
}
```

though it's hard to know if this is a complete definition since documentation regarding the API seems to be outdated

**Review by davkal — APPROVED** [review]
Tested with Prometheus and Graphite. LGTM

**Comment by torkelo** [comment]
wait with merge, I have some things I need to test and make sure there is no breaking change here. So many changes to dashboardSrv etc.

File: public/app/features/dashboard/services/DashboardSrv.ts [hunk 11]
```diff
@@ -152,17 +157,24 @@ export class DashboardSrv {
...
-  saveDashboard(options?: { overwrite?: any; folderId?: any; makeEditable?: any }, clone?: DashboardModel) {
+  saveDashboard(
+    clone?: DashboardModel,
+    {
+      makeEditable = false,
+      folderId,
+      overwrite = false,
+      message,
+    }: DashboardSaveOptions & { makeEditable?: boolean } = {}
+  ) {
```
Comment (torkelo) [inline public/app/features/dashboard/services/DashboardSrv.ts]: this function definition feels a bit messy and hard to read. Wonder if there is an easier way to wite this.

**Review by torkelo — COMMENTED** [review]
This is looking good now. Feel like the complexity is a bit lower now that this is handled in the dashboard init phase instead of in the UI component. 

* When I am using Test data datasource (gdev-testdata) I get no option to return with changes

File: public/app/features/dashboard/panel_editor/QueriesTab.tsx [hunk 12]
```diff
@@ -273,3 +273,5 @@ export class QueriesTab extends PureComponent<Props, State> {
     );
   }
 }
+
+export default QueriesTab;
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/dashboard/panel_editor/QueriesTab.tsx]: Fair enough, will change

File: public/app/features/dashboard/state/initDashboard.ts [hunk 13]
```diff
@@ -198,6 +200,24 @@ export function initDashboard(args: InitDashboardArgs): ThunkResult<void> {
...
+        dashboard.panels[panelArrId].targets = left.queries.map((query: DataQuery & { context?: string }) => {
+          delete query.context;
+          delete query.key;
+          return query;
+        });
+      }
+      dashboard.startRefresh();
+      // Force-reset explore so that on subsequent dashboard loads we aren't
+      // taking the modified queries from explore again.
+      dispatch(resetExploreAction({ force: true }));
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/dashboard/state/initDashboard.ts]: Agree that this makes more sense. Will do

File: public/app/features/explore/ExploreToolbar.tsx [hunk 14]
```diff
@@ -112,6 +122,31 @@ export class UnConnectedExploreToolbar extends PureComponent<Props, {}> {
...
+    if (!withChanges) {
+      this.props.clearOrigin();
+    }
+
+    const dashViewOptions = {
+      fullscreen: withChanges || dash.meta.fullscreen,
+      edit: withChanges || dash.meta.isEditing,
+    };
+
+    getLocationSrv().update({
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/explore/ExploreToolbar.tsx]: Cool, will change

File: public/app/features/dashboard/services/DashboardSrv.ts (hunk 11 above)
Comment (kaydelaney) [reply public/app/features/dashboard/services/DashboardSrv.ts]: I'll see what I can do

Comment (kaydelaney) [reply public/app/features/dashboard/services/DashboardSrv.ts]: > This is looking good now. Feel like the complexity is a bit lower now that this is handled in the dashboard init phase instead of in the UI component.
> 
>     * When I am using Test data datasource (gdev-testdata) I get no option to return with changes

@torkelo If the dashboard the panel is on is provisioned or otherwise can't be saved then the option to return with changes is disabled. Can undo if you'd prefer though.

Comment (torkelo) [reply public/app/features/dashboard/services/DashboardSrv.ts]: ah, that explains it. I think the provisioned check is not needed as you can edit provisioned dashboards in UI and then when you try to save you will see the json

File: public/app/features/dashboard/state/initDashboard.test.ts [hunk 15]
```diff
@@ -101,6 +108,19 @@ describeInitScenario('Initializing new dashboard', ctx => {
   ctx.setup(() => {
     ctx.storeState.user.orgId = 12;
     ctx.args.routeInfo = DashboardRouteInfo.New;
+    ctx.storeState.explore.left.originPanelId = 2;
```
Comment (torkelo) [inline public/app/features/dashboard/state/initDashboard.test.ts]: would like to move this to its own scenario, and not have it be part of the "New dashboard" scenario.

**Review by torkelo — APPROVED** [review]
Looks good! just one minor thing  with the initDashboard test ..

File: public/app/features/dashboard/state/initDashboard.test.ts (hunk 15 above)
Comment (kaydelaney) [reply public/app/features/dashboard/state/initDashboard.test.ts]: Fair enough, had some trouble trying to do this initially but I can give it another go 😅

**Review by davkal — COMMENTED** [review]
Works well now.

Found another small issue: after returning with edits, the dashboard is not fully aware that it's been modified. Repro:

1. Jump to explore from a panel (i used a graphite DS)
2. do an edit, and reutrn with edits, edits are preserved
3. now move out of edit mode, the edit is still preserved
4. move away from the dashboard (e.g., click on Explore)

Result: no warning modal asking to save the modification.

**Comment by kaydelaney** [comment]
> Works well now.
> 
> Found another small issue: after returning with edits, the dashboard is not fully aware that it's been modified. Repro:
> 
>     1. Jump to explore from a panel (i used a graphite DS)
> 
>     2. do an edit, and reutrn with edits, edits are preserved
> 
>     3. now move out of edit mode, the edit is still preserved
> 
>     4. move away from the dashboard (e.g., click on Explore)
> 
> 
> Result: no warning modal asking to save the modification.

Ah, nice catch. Let me look into this

**Review by davkal — APPROVED** [review]
LGTM

File: public/app/features/dashboard/services/DashboardSrv.ts [hunk 16]
```diff
@@ -37,54 +44,53 @@ export class DashboardSrv {
...
   };
 
   handleSaveDashboardError(
     clone: any,
-    options: { overwrite?: any },
+    options: DashboardSaveOptions,
     err: { data: { status: string; message: any }; isHandled: boolean }
   ) {
-    options = options || {};
-    options.overwrite = true;
```
Comment (torkelo) [inline public/app/features/dashboard/services/DashboardSrv.ts]: by removing options.overwrite = true none of these overwrite scenarios work
